apps/rush-lib/src/logic/UpgradeRushSelf.ts

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import * as semver from 'semver';
+import * as child_process from 'child_process';
+import colors from 'colors/safe';
+import {
+  ConsoleTerminalProvider,
+  Executable,
+  FileSystem,
+  Import,
+  InternalError,
+  JsonFile,
+  JsonObject,
+  Terminal
+} from '@rushstack/node-core-library';
+
+import { RushConfiguration } from '../api/RushConfiguration';
+import type * as inquirerTypes from 'inquirer';
+import { Autoinstaller } from './Autoinstaller';
+const inquirer: typeof inquirerTypes = Import.lazy('inquirer', require);
+
+export interface IUpgradeRushSelfOptions {
+  rushConfiguration: RushConfiguration;
+  isDebug: boolean;
+}
+
+export interface IUpgradeResult {
+  needRushUpdate: boolean;
+}
+
+export class UpgradeRushSelf {
+  private readonly _rushConfiguration: RushConfiguration;
+  private readonly _terminal: Terminal;
+  private readonly _isDebug: boolean = false;
+
+  public constructor(options: IUpgradeRushSelfOptions) {
+    this._rushConfiguration = options.rushConfiguration;
+    this._isDebug = options.isDebug;
+    this._terminal = new Terminal(
+      new ConsoleTerminalProvider({
+        verboseEnabled: options.isDebug
+      })
+    );
+  }
+
+  public getAvaiableRushVersions(): string[] {
+    this._terminal.writeLine('Fetching rush versions');
+
+    const { versions }: { versions: string[] } = this._npmView('@microsoft/rush');
+    if (!Array.isArray(versions)) {
+      throw new Error('Get @microsoft/rush versions incorrectly');
+    }
+
+    return versions;
+  }
+
+  public async upgrade(): Promise<IUpgradeResult> {
+    const versions: string[] = this.getAvaiableRushVersions();
+
+    const promptModule: inquirerTypes.PromptModule = inquirer.createPromptModule();
+    const { version } = await promptModule({
+      type: 'list',
+      name: 'version',
+      message: `Select a version to upgrade`,
+      choices: versions.sort((a, b) => semver.rcompare(a, b))
+    });
+
+    // update rush.json
+    const rushJsonFile: string = this._rushConfiguration.rushJsonFile;
+    const json: JsonObject = JsonFile.load(rushJsonFile);
+    if (json.rushVersion === version) {
+      console.log();
+      console.log(colors.yellow(`upgrading same rush version: ${version}`));
+      return {
+        needRushUpdate: false
+      };
+    }
+    json.rushVersion = version;
+    JsonFile.save(json, rushJsonFile, { updateExistingFile: true, onlyIfChanged: true });
+    if (this._isDebug) {
+      console.log(colors.gray(`upgrade rushVersion in rush.json to ${version}`));
+    }
+
+    const dependencies: Record<string, string> = this._npmView(`@microsoft/rush@${version}`, 'dependencies');
```

Comment:
This part may step on some toes. I think we should perhaps only update `@microsoft/rush-lib` dependencies in projects' `package.json`s.